### ISSUE TITLE
Make AddUninstallTarget FetchContent / add_subdirectory friendly

### ISSUE THREAD

**Issue by traversaro** [robotology/ycm#169]
Similar to what we discussed in https://github.com/robotology/ycm/issues/162#issue-349119219, also `AddUninstallTarget` (or any other module that adds custom "global" targets) is not really compatible with the FetchContent / add_subdirectory pattern (see https://github.com/robotology/idyntree/issues/370#issuecomment-420705347). 

I think this can be elegantly addressed by adding the uninstallation code in a custom target `${PROJECT_NAME}-uninstall`, and adding this target as a dependency to the `uninstall` target, eventually creating the `uninstall` target if it does not already exist. 

Something like (warning: this was not actually tested):
~~~cmake
add_custom_target(${PROJECT_NAME}-${_uninstall} COMMAND "${CMAKE_COMMAND}" -P "${_filename}")
set_property(TARGET ${PROJECT_NAME}-${_uninstall} PROPERTY FOLDER "CMakePredefinedTargets")

if(NOT TARGET ${_uninstall})
  add_custom_target(${_uninstall})
endif()
add_dependencies(${_uninstall}  ${PROJECT_NAME}-${_uninstall})
~~~

cc @drdanz @claudiofantacci  @ahoarau

**Comment by traversaro** [comment]
I just realized that this does not make any sense, as the `install_manifest.txt` is only one for project. If there is already an uninstall target, the only thing we can do is hope that it does the right thing, and we do not add any additional uninstall target.

**Comment by traversaro** [comment]
We just need to have a: 
~~~cmake
if(TARGET ${_uninstall})
  return()
endif()
~~~
somewhere, ideally before the `file(WRITE` to avoid polluting the build with confusing files.

**Comment by traversaro** [comment]
Indeed the latest version of the CMake's `uninstall`s FAQ's entry is doing exactly this:
https://gitlab.kitware.com/cmake/community/wikis/FAQ#can-i-do-make-uninstall-with-cmake

**Comment by claudiofantacci** [comment]
Thank @traversaro for raising this valid point.

Indeed, as long as the `install_manifest` is correctly populated, the uninstall target is well-formed. However there shall be only one `uninstall` target. I'm sure that CMake complains when it finds the same target multiple times and allows only the first one (just try to add the uninstall target with YCM and ICUBcontrib and you'll see it).

Adding
```cmake
if(TARGET ${_uninstall})
  return()
endif()
```
would be the cleanest solution and we will also avoid CMake warnings.

**Comment by drdanz** [comment]
I just realized that these lines are supposed to do that, but that doesn't work in subdirectories.

https://github.com/robotology/ycm/blob/7c59d08b40c5e52cd1b4f5c6339a6f9b28b24cb9/modules/AddUninstallTarget.cmake#L30-L33

**Comment by drdanz** [comment]
In #237 I added a double check: `${CMAKE_CURRENT_SOURCE_DIR}` must be equal to `${CMAKE_SOURCE_DIR}`, and the `uninstall` target must not exist.
Please let me know what you think about this...

**Comment by traversaro** [comment]
I am not sure, I think including `AddUninstallTarget` in a subdirectory of a project should be allowed.

**Comment by drdanz** [comment]
That's what I thought at the beginning, but then I realised that it looks for the `install_manifest.txt` file in `CMAKE_CURRENT_BINARY_DIR`, that means it never worked. Now it is explicit...

**Comment by traversaro** [comment]
Ah, good point!

**Comment by traversaro** [comment]
Perhaps it should check for it in `CMAKE_BINARY_DIR`, as suggested in CMake FAQs: https://gitlab.kitware.com/cmake/community/wikis/FAQ#can-i-do-make-uninstall-with-cmake .

**Comment by drdanz** [comment]
Maybe, but in this case the command will not allow a build including the project using FetchContent to define the `uninstall` target, unless this is done *before* adding the project subdirectory.
This is not a problem if the importing project uses AddUninstallTarget, but it is a problem if someone wants to create a different/custom target.
Forcing the user to include it in the main `CMakeLists.txt` makes it possible, and does not break compatibility, since it has always been like that, since we introduced the module in YCM.

BTW, I noticed that `CTestTargets.cmake` (included by `CTest.cmake`) does this:

```cmake
define_property(GLOBAL PROPERTY CTEST_TARGETS_ADDED
  BRIEF_DOCS "Internal property used by CTestTargets module."
  FULL_DOCS "Set by the CTestTargets module to track addition of testing targets."
  )
get_property(_CTEST_TARGETS_ADDED GLOBAL PROPERTY CTEST_TARGETS_ADDED)
if(NOT _CTEST_TARGETS_ADDED)
  set_property(GLOBAL PROPERTY CTEST_TARGETS_ADDED 1)
  # [...]
```

`CTest.cmake` has the same behaviour, it creates some targets, and it is mandatory to include it in the main `CMakeLists.txt`

**Comment by traversaro** [comment]
> Forcing the user to include it in the main `CMakeLists.txt` makes it possible, and does not break compatibility, since it has always been like that, since we introduced the module in YCM.

Indeed, this is a nice case of a problem that quickly become a feature!

**Comment by drdanz** [comment]
BTW, it has been like this since the very beginning of YARP, see robotology/yarp@3faaba0974833d719ffe91768bb6c9cc13ba5032
Probably it was copied from some FAQ/Wiki that, at the time, was using `CMAKE_CURRENT_BINARY_DIR`

**Comment by drdanz** [comment]
Found: https://cmake.org/Wiki/index.php?title=CMake_FAQ&oldid=36041#Can_I_do_.22make_uninstall.22_with_CMake.3F :rofl:

**Comment by drdanz** [comment]
Fixed in #237
